exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterSink.java

```diff
@@ -74,32 +70,43 @@
 
   private SendingAccountor sendingAccountor;
 
+  private  AsyncAggregateWorker asyncAggregateWorker;
+
+  private AtomicBoolean running = new AtomicBoolean(true);
+
   private static final Logger logger = LoggerFactory.getLogger(RuntimeFilterSink.class);
 
 
   public RuntimeFilterSink(DrillbitContext drillbitContext, SendingAccountor sendingAccountor)
   {
     this.drillbitContext = drillbitContext;
     this.sendingAccountor = sendingAccountor;
-    AsyncAggregateWorker asyncAggregateWorker = new AsyncAggregateWorker();
-    drillbitContext.getExecutor().submit(asyncAggregateWorker);
   }
 
   public void add(RuntimeFilterWritable runtimeFilterWritable)
   {
+    if (asyncAggregateWorker == null) {
+      asyncAggregateWorker = new AsyncAggregateWorker();
+      drillbitContext.getExecutor().submit(asyncAggregateWorker);
+    }
     runtimeFilterWritable.retainBuffers(1);
     int joinMjId = runtimeFilterWritable.getRuntimeFilterBDef().getMajorFragmentId();
     if (joinMjId2Stopwatch.get(joinMjId) == null) {
       Stopwatch stopwatch = Stopwatch.createStarted();
       joinMjId2Stopwatch.put(joinMjId, stopwatch);
     }
-    queueLock.lock();
-    try {
+    synchronized (rfQueue) {
       rfQueue.add(runtimeFilterWritable);
```

Comment:
should check `running` status before adding in queue. In cases when running is false the thread should `close` the received filter